File: csharp/ToolGood.Algorithm2/LitJson/JsonMapper.cs
namespace ToolGood.Algorithm.LitJson
{
   sealed class JsonMapper
    {

        #region Private Methods

        private static IJsonWrapper ReadValue(JsonReader reader)
        {
            reader.Read();

            if (reader.Token == JsonToken.ArrayEnd) return null;

            IJsonWrapper instance = new JsonData();

            if (reader.Token == JsonToken.String) {
                instance.SetString((string)reader.Value);
                return instance;
            }

            if (reader.Token == JsonToken.Double) {
                instance.SetDouble((decimal)reader.Value);
                return instance;
            }

            if (reader.Token == JsonToken.Boolean) {
                instance.SetBoolean((bool)reader.Value);
                return instance;
            }
            if (reader.Token == JsonToken.Null) {
                instance.SetNull();
                return instance;
            }


            if (reader.Token == JsonToken.ArrayStart) {
                instance.SetJsonType(JsonType.Array);

                while (true) {
                    IJsonWrapper item = ReadValue(reader);
                    if (item == null && reader.Token == JsonToken.ArrayEnd) break;
                    instance.Add(item);
                }
            } else if (reader.Token == JsonToken.ObjectStart) {
                instance.SetJsonType(JsonType.Object);

                while (true) {
                    reader.Read();

                    if (reader.Token == JsonToken.ObjectEnd) break;

                    string property = (string)reader.Value;
                    instance.Set(property, ReadValue(reader));
                }

            }

            return instance;
        }

        #endregion



        public static JsonData ToObject(string json)
        {
            JsonReader reader = new JsonReader(json);
            return ReadValue(reader) as JsonData;
        }

    }
}

File: JavaScript/src/LitJson/JsonMapper.js
import { JsonData } from './JsonData.js';
import { JsonReader,JsonToken } from './JsonReader.js';
import { JsonType } from './JsonType.js';

class JsonMapper {
    static readValue(reader) {
        reader.read();

        if (reader.Token === JsonToken.ArrayEnd) return null;

        let instance = new JsonData();

        if (reader.Token === JsonToken.String) {
            instance.setString(reader.Value);
            return instance;
        }

        if (reader.Token === JsonToken.Double) {
            instance.setDouble(reader.Value);
            return instance;
        }

        if (reader.Token === JsonToken.Boolean) {
            instance.setBoolean(reader.Value);
            return instance;
        }
        if (reader.Token === JsonToken.Null) {
            instance.setNull();
            return instance;
        }

        if (reader.Token === JsonToken.ArrayStart) {
            instance.setJsonType(JsonType.Array);

            while (true) {
                let item = JsonMapper.readValue(reader);
                if (item === null && reader.Token === JsonToken.ArrayEnd) break;
                instance.add(item);
            }
        } else if (reader.Token === JsonToken.ObjectStart) {
            instance.setJsonType(JsonType.Object);

            while (true) {
                reader.read();

                if (reader.Token === JsonToken.ObjectEnd) break;

                let property = reader.Value;
                instance.set(property, JsonMapper.readValue(reader));
            }
        }

        return instance;
    }

    static toObject(json) {
        let reader = new JsonReader(json);
        return JsonMapper.readValue(reader);
    }
}

export { JsonMapper };